frameit documentation should describe all parameters from the Framefile

funnel20 opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
<!-- Please describe in detail how you tested your changes. -->
This is the start to improve the Frameit docu, as indicated in issue #11647.

### Description
<!-- Describe your changes in detail -->
This PR is WIP, as it will be changed on the fly by several authors, like @janpio

funnel20 committed:
This is the start to improve the Frameit docu, as indicated in issue https://github.com/fastlane/fastlane/issues/11647

janpio committed:
added first notes to improve docs

janpio commented:
I went through the document and added the first few notes and TODOs to the current text.

Could you maybe add the missing parameters in a rudimentary form next @funnel20 ? Text or order doesn't really matter, just that it is there. Also linking to the relevant source code here in a comment for what you added would be nice to be able to quickly read about it.

funnel20 commented:
The file that we are editing seems be nested in https://docs.fastlane.tools/actions/frameit/.
At the bottom is a table **Parameters** that describes all the parameters for the Frameit command line options.

I would prefer a similar table **Framefile Parameters** to list all existing and missing parameters. As a table is much more easy to read, than a bunch of text with nested parameter names.
Do you agree, than I'll construct the table?

funnel20 commented:
How can I visually check the *.md file?
From the fastlane pages it loads fine in my Safari browser.
But Safari fails to open it from my local fork.

janpio commented:
> Do you agree, than I'll construct the table?

Yes, that's exactly what I meant by "TODO parameters in usual table instead of normal text".

> How can I visually check the *.md file?

1. Use Visual Code locally, that can render a preview of an .md file.
2. Edit the file in Github UI, which has a preview of .md files.
3. Use any other Markdown editor to open end edit the file, most of them have a preview method.

Not that they all probably won't show the images/screenshots included in the document.

funnel20 commented:
Installed Visual Studio Code, works as a charm. Thanks for the tip.

I don't have much time this week, but I try to make some progress.

janpio commented:
We are not in a hurry, so take your time 👍

funnel20 committed:
Here are 2 different table options to specify the keys:
1) a key-description representation of the existing text
2) a table that describes the JSON hierarchy with sub keys

Which one is preferred?

funnel20 commented:
Please see my latest commit, which of the 2 tables is preferred?

janpio commented:
Representing JSON in a table is too confusing I think. 

Better describe the properties of `default` and `data` in two separate tables after introducing the overall structure first maybe?

```
{
  "device_frame_version": "latest",
  "default": {
    ...
  },
  "data": [
     ...
  ]
}
```

Then finish with the full example that shows all complexities.

funnel20 committed:
Used two separate tables after introducing the overall JSON to describe the keys of `default` and `data`.

funnel20 commented:
See my latest commit.

janpio reviewed (changes requested):
Not nearly ready to merge. (Making sure nobody merges this by accident)

janpio commented on fastlane/lib/fastlane/actions/docs/frame_screenshots.md:
```diff
@@ -68,18 +81,66 @@ To download the latest frames
+### General parameters
+The general parameters are defined in the `default` key and can be:
+
+| Key | Description | Default value |
```
missing:

funnel20 commented:
This was just an update of our first TODO; how to represent the parameters.

janpio committed:
add background, reword advanced intro

janpio commented on fastlane/lib/fastlane/actions/docs/frame_screenshots.md:
```diff
@@ -68,18 +81,70 @@ To download the latest frames
+
+The general parameters are defined in the `default` key and can be:
+
+| Key | Description | Default value |
```
missing: `offset` for Mac https://docs.fastlane.tools/actions/frameit/#example

funnel20 replied:
Shall we make an extra table with the Mac specific parameters?

janpio replied:
Sounds like a good iea, yeah.

janpio commented:
Your tables look good I think, structure makes a lot more sense that way for someone new trying to learn I think. Do you agree?

funnel20 commented:
Hmm, I didn't see your commit, so destroyed my branch because I made a commit.
Back to the study books.... Which probably means that I give up

janpio commented:
Huh? Destroy how?
You can either merge (another topic to learn about...) or just check out the most recent version with my commit again in a new folder and apply your last changes again.

funnel20 committed:
Test if I can get local anchor links working.

funnel20 commented:
I had merging conflicts and tons of errors in Sourcetree, so deleted the whole local copy and checked out again.

funnel20 committed:
* Described `background`
* Added anchor links

funnel20 commented:
@janpio What do you want to do with my TODOs about `color` and `font` in table Framefile keyword and title parameters?

janpio commented on fastlane/lib/fastlane/actions/docs/frame_screenshots.md:
```diff
@@ -135,9 +135,10 @@ These are defined in the `data` key. This is an array with the following keys fo
 | `title` | Similar use as in `default`, except that parameter `text` can be used here because it is screenshot specific. |
 
-#### Framefile `keyword` and `title` parameters
+#### <a name="keyword-and-title-parameters"></a>Framefile `keyword` and `title` parameters
```
This is usually autogenerated when deploying the site. See https://docs.fastlane.tools/actions/frameit/#titles-and-background-optional as an example

funnel20 replied:
I prefer to have this under control, because we're referring to it within this document.
When this section title will be renamed later (by someone else), and the site automatically generates a new anchor, the existing reference to the changed anchor will no longer work.

janpio commented:
> What do you want to do with my TODOs about color and font in table Framefile keyword and title parameters?

Good question. These sound pretty complicated...

funnel20 commented:
`color`
The frameit input validator only allows a hex input for `color`, so that problem is solved.
Here is the Terminal output when defining: `"color": "red"`

> Invalid color 'red'. Must be valid Hex #123123

`font`
Referring that the default value is system dependant, should be enough to convince user to explicitly specify a font.

So this is my proposal:

> The font family for the text. Specify the (relative) path to the font file (e.g. a TrueType Font). | The default MiniMagick font, which is system dependent.

janpio commented:
Sounds good to me!

funnel20 committed:
- Added missing parameters `padding`, `interline_spacing` and `font_scale_factor`
- Completed descriptions for `color` and `font`
- Combined NOTES in section "`.strings` files"
- Restructured MAC section and extended descriptions
- Applied consistent header sizes
- Removed original parameter descriptions below Example.
- Removed completed TODO comments

funnel20 commented:
@janpio I made several changes, see my last commit. Unfortunately it shows a conflict. Is this something that you can resolve?

Can you please review the changes in the section MAC? I haven't any experience with Frameit for MAC, so I changed the text according to what I expect from the MAC example and the source code.

funnel20 commented:
I've intentionally left the `offset` out of the tables for Advanced Usage, as it's an extremely confusion parameter. The concept of combining both `offset` and `titleHeight` within an object named `offset` (again) is not clear to me. Especially since there is no `keywordHeight`, and the code also mentions a `width` inside the top `offset` object, i.e. `offset['width']`, see: https://github.com/fastlane/fastlane/blob/181a4c140ceca56c0ebf89bcdfa452527968483b/frameit/lib/frameit/editor.rb#L215

janpio committed:
Merge branch 'master' into improve-fameit-docu

janpio commented:
Took care of the merge conflict. This happened because someone else modified the same file, and Git couldn't figure out how to merge these changes with ours her because we moved stuff around and also changed some sentences so Git doesn't recognize the old version and the new version as the same line - and conflicts.

Will take a look at your changes later.

funnel20 commented:
The only thing that I can find about `device_frame_version` is that it can be a time stamp:
https://github.com/fastlane/fastlane/blob/181a4c140ceca56c0ebf89bcdfa452527968483b/frameit/lib/frameit/module.rb#L14-L15

But where to find a list of valid time stamps?

funnel20 commented:
Thanks for merging. Were these changes done on the same branch? I double-checked in SourceTree and in the commit list of this PR, it seems that my local copy was up-to-date.

funnel20 committed:
Corrected typo in `padding`

janpio commented:
No, the conflict shown here was a preview if someone would have merged into `master`. In `master` some changes happened that would have prevented a merge. My "Resolve conflict" effectively just told Git, that whatever is is `master`, our version here is better and should be used :)

funnel20 commented:
@janpio Do you have some time to review my changes and answer the last 2 questions?
We're finally there, so I'd like to complete this.

funnel20 commented:
@janpio Can you please review this or assign someone else, so we can complete this PR?

janpio committed:
clean up: remove todo and other stuff, add android upload

funnel20 commented:
@janpio Thx!!

revolter commented on fastlane/lib/fastlane/actions/docs/frame_screenshots.md:
```diff
@@ -2,7 +2,7 @@
 </p>
 
-###### Quickly put your screenshots into the right device frames
+###### Easily put your screenshots into the right device frames
```
I think that @KrauseFx should decide if it's ok to change this tagline.

janpio replied:
This actually was just a change @funnel20 made to start the PR, can just be reverted.

funnel20 replied:
Please revert. The whole idea of this PR is to bring this docu in line with the current code implementation and this word is not related.

revolter commented on fastlane/lib/fastlane/actions/docs/frame_screenshots.md:
```diff
@@ -21,19 +21,30 @@ _frameit_ allows you to put a gorgeous device frame around your iOS and macOS sc
-### Results
+## Advanced Features
+
+- Put framed screenshot on colored background, define padding
```
Should `define padding` be on its own line? Maybe `- Define custom padding`?

funnel20 replied:
I only focussed on adding the missing parameters, not on these kind of optimisations. I prefer to leave this as is.

revolter commented on fastlane/lib/fastlane/actions/docs/frame_screenshots.md:
```diff
@@ -21,19 +21,30 @@ _frameit_ allows you to put a gorgeous device frame around your iOS and macOS sc
+  - keyword + text
+  - choose text font and color
+  - multi line text
+  - "intelligent" positioning of text that always looks good(ish)
```
Personally, I'd change these to:

```
- Add text above or under the framed screenshot, allowing you to:
  - Add a keyword along with the text (not sure about this)
  - Choose the font and color
  - Make it multilined
  - Automatic positioning for almost perfect looks
```

janpio replied:
> Add a keyword along with the text (not sure about this)

Nah, "keyword" and "text" are referring to the "text fields" being in use here. Pretty technical (and bad) terms but defined in code.

revolter replied:
What about the rest of the changes? Including uppercased items.

janpio replied:
I don't care - but sounds good.

funnel20 replied:
Some changes that better reflect the frameit terms:

- Add text above or under the framed screenshot, allowing you to:
  - This can be a keyword with optional title
  - Choose the font and color
  - frameit supports multi-line text
  - Automatic positioning for almost perfect looks

revolter commented on fastlane/lib/fastlane/actions/docs/frame_screenshots.md:
```diff
@@ -70,16 +83,69 @@ To download the latest frames
+### Text and Background
 
-With _frameit_ 2.0 you are now able to add a custom background, title and text colors to your screenshots.
+With _frameit_ it's possible to add a custom background and text below or above the framed screenshots in fonts and colors you define.
```
I'd use commas to separate what appears to be a list of features: `With _frameit_, it's possible to add a custom background and text, below or above the framed screenshots, in fonts and colors you define.`

janpio replied:
First one: Nope, as without the text after the comma it would not be a complete sentence any more.
Second and third one: Could do, if "below or above the framed screenshots" should be an addition that could be left off. Here I would be okay with them.

revolter replied:
For the first one, Grammarly also marked it as an error, but if it makes sense without it too, I have no problem with it.

![screen shot 2018-03-19 at 22 39 44](https://user-images.githubusercontent.com/5748627/37621077-9ed7ee28-2bc6-11e8-99d6-fc813970a11f.png)

janpio replied:
English is weird.

funnel20 replied:
What about:
With _frameit_ it's possible to add a custom background and text below or above the framed screenshots. You can define the text fonts and colors.

revolter commented on fastlane/lib/fastlane/actions/docs/frame_screenshots.md:
```diff
@@ -70,16 +83,69 @@ To download the latest frames
-Use it to define the general information:
+The general parameters are defined in the `default` key and can be:
 
+| Key | Description | Default value |
```
This should not be here. The _frameit_'s source code should be changed (create `ConfigItem`s for these missing options) so the docs would be correctly generated.

janpio replied:
see my comment below on the main PR comments thread

revolter replied:
I see now, didn't know the difference between _frameit_ and the rest of the tools. Knowing this, you should disregard this review comment, at least for now.

revolter commented:
Is manually adding the missing options in the `frame_screenshots.md` file the only way? I don't know the background of the issue or _frameit_'s source code, though I can take a look into it.

janpio commented:
These are not really "options" in the classical sense, but structure of a configuration file. `frameit` works pretty differently from the rest of the stuff it seems, probably also why this wasn't really documented before.

funnel20 commented:
@revolter Thanks for your feedback. I have added some comments and suggestions.
Please note the title of this PR. Our main purpose for now is to have all parameters from the JSON configuration properly documented.
So I'd prefer to create a new issue/PR for anything else which is faulty or obsolete, so we can at least commit this important update.

funnel20 commented:
@revolter thanks for approving this
